### ISSUE TITLE
Resources have not been updated after the overridepolicyis updated.

### ISSUE THREAD

**Issue by jrkeen** [karmada-io/karmada#507]
**What happened**:
1. Current environment.
```
## OverridePolicy:
apiVersion: policy.karmada.io/v1alpha1
kind: OverridePolicy
metadata:
  creationTimestamp: "2021-07-08T07:37:50Z"
  generation: 1
  name: nginx3333
  namespace: test
spec:
  overriders:
    plaintext:
    - operator: add
      path: /metadata/annotations
      value:
        aa: bb
  resourceSelectors:
  - apiVersion: apps/v1
    kind: Deployment
    name: nginx3333
    namespace: test
  targetCluster:
    clusterNames:
    - host

## Deployment has updated in host:
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    aa: bb
  creationTimestamp: "2021-07-08T06:13:18Z"
  generation: 2
  labels:
    app: nginx3333
  name: nginx3333
  namespace: test
```

2. Updated overridepolicy in Karmada.
```
## OverridePolicy:
apiVersion: policy.karmada.io/v1alpha1
kind: OverridePolicy
metadata:
  creationTimestamp: "2021-07-08T07:37:50Z"
  generation: 1
  name: nginx3333
  namespace: test
spec:
  overriders:
    plaintext:
    - operator: add
      path: /metadata/annotations
      value:                                           <--------- value has updated
        bb: cc
  resourceSelectors:
  - apiVersion: apps/v1
    kind: Deployment
    name: nginx3333
    namespace: test
  targetCluster:
    clusterNames:
    - host
 
## Deployment had not updated in host:
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    aa: bb
  creationTimestamp: "2021-07-08T06:13:18Z"
  generation: 2
  labels:
    app: nginx3333
  name: nginx3333
  namespace: test
```

**Anything else we need to know?**:
Is this the design of karmada itself ?
I think the deployment should update when updated overridepolicy.

**Environment**:
- Karmada version: latest

**Comment by RainbowMango** [comment]
Sorry for let this sit so long. @XiShanYongYe-Chang can you help to confirm this?

**Comment by XiShanYongYe-Chang** [comment]
> Sorry for let this sit so long. @XiShanYongYe-Chang can you help to confirm this?

Ok, let me have a look.

**Comment by XiShanYongYe-Chang** [comment]
Hi, @jrkeen it is a code bug.

When overridepolicy has changed, it will be work on related resource template, and `execution-controller` will update it in:

https://github.com/karmada-io/karmada/blob/764cbbc87641575164b3e2a5cd9a1d192d5c0022/pkg/util/objectwatcher/objectwatcher.go#L100

Here will retain some fields of `cluster object` to `desire object`:

https://github.com/karmada-io/karmada/blob/764cbbc87641575164b3e2a5cd9a1d192d5c0022/pkg/util/objectwatcher/objectwatcher.go#L113-L117

In `RetainClusterFields` method, we will override `desire object`'s Annotation with `cluster object`, resulting in an error which you mentioned above:

https://github.com/karmada-io/karmada/blob/764cbbc87641575164b3e2a5cd9a1d192d5c0022/pkg/util/objectwatcher/retain.go#L33

**Comment by XiShanYongYe-Chang** [comment]
We can merge the annotation from `cluster object` and `desire object`.

There is still a flaw after the merger, but I think it's acceptable:

After updated overridepolicy in Karmada, deployment will update in member cluster. There will add a new annotation key-value `bb: cc` in deployment, but the old annotation key-value `aa: bb` still exist. Just like this:

```
## Deployment had not updated in host:
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    aa: bb
    bb: cc
    ...         # other annotation key-value added by member cluster 
  creationTimestamp: "2021-07-08T06:13:18Z"
  generation: 2
  labels:
    app: nginx3333
  name: nginx3333
  namespace: test
```

**Comment by jrkeen** [comment]
> Hi, @jrkeen it is a code bug.
> 
> When overridepolicy has changed, it will be work on related resource template, and `execution-controller` will update it in:
> 
> https://github.com/karmada-io/karmada/blob/764cbbc87641575164b3e2a5cd9a1d192d5c0022/pkg/util/objectwatcher/objectwatcher.go#L100
> 
> Here will retain some fields of `cluster object` to `desire object`:
> 
> https://github.com/karmada-io/karmada/blob/764cbbc87641575164b3e2a5cd9a1d192d5c0022/pkg/util/objectwatcher/objectwatcher.go#L113-L117
> 
> In `RetainClusterFields` method, we will override `desire object`'s Annotation with `cluster object`, resulting in an error which you mentioned above:
> 
> https://github.com/karmada-io/karmada/blob/764cbbc87641575164b3e2a5cd9a1d192d5c0022/pkg/util/objectwatcher/retain.go#L33

Thanks.

**Comment by XiShanYongYe-Chang** [comment]
@jrkeen , the fix code #531 has been merged,  you can ues the newest code have a try.

**Comment by jrkeen** [comment]
> @jrkeen , the fix code #531 has been merged, you can ues the newest code have a try.

OK, I will do it.
